src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -0,0 +1,404 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    public class PatternMatchingTests_ListPatterns : PatternMatchingTestBase
+    {
+        [Fact]
+        public void ListPattern()
+        {
+            string testMethod(string type) =>
+@"static bool Test(" + type + @" input)
+{
+    switch (input)
+    {
+        case [0]:
+        case {_}:
+          return true;
+        case {var first, ..var others, var last} when first == last:
+          return Test(others);
+        default:
+          return false;
+    }
+}";
+            var source = @"
+using System;
+public class X
+{
+    " + testMethod("Span<char>") + @"
+    " + testMethod("char[]") + @"
+    " + testMethod("string") + @"
+    static void Check(int num)
+    {
+        Console.Write(Test((string)num.ToString()) ? 1 : 0);
+        Console.Write(Test((char[])num.ToString().ToCharArray()) ? 1 : 0);
+        Console.Write(Test((Span<char>)num.ToString().ToCharArray()) ? 1 : 0);
+        Console.WriteLine();
+    }
+    public static void Main()
+    {
+        Check(1);
+        Check(11);
+        Check(12);
+        Check(123);
+        Check(121);
+        Check(1221);
+        Check(1222);
+    }
+}
+" + TestSources.GetSubArray;
+            var compilation = CreateCompilationWithIndexAndRangeAndSpan(source, parseOptions: TestOptions.RegularWithListPatterns, options: TestOptions.ReleaseExe);
+            compilation.VerifyDiagnostics();
+            var expectedOutput = @"
+111
+111
+000
+000
+111
+111
+000";
+            var verifier = CompileAndVerify(compilation, expectedOutput: expectedOutput);
+            verifier.VerifyIL("X.Test(System.Span<char>)", @"
+{
+  // Code size       76 (0x4c)
+  .maxstack  4
+  .locals init (char V_0, //first
+                System.Span<char> V_1, //others
+                char V_2, //last
+                System.Span<char> V_3,
+                int V_4)
+  IL_0000:  ldarg.0
+  IL_0001:  stloc.3
+  IL_0002:  ldloca.s   V_3
+  IL_0004:  call       ""int System.Span<char>.Length.get""
```

Comment:
We should probably also emit a null check on the input value before accessing the `Length` property.

Reply:
We do. It's just omitted for a Span struct.

Reply:
Ah, indeed. I see it for cases below. Thanks